File: 220916/Java/KeyPad.java
public class KeyPad {
  public String solution(int[] numbers, String handUse) {
    String result = "";

    int[] leftFingerPosition = new int[]{3, 0};
    int[] rightFingerPosition = new int[]{3, 2};

    for (int number : numbers) {
      String handUsed
          = selectFinger(number, leftFingerPosition, rightFingerPosition, handUse);

      result += handUsed;

      if (handUsed.equals("L")) {
        leftFingerPosition = moveFingerPosition(number);
      }
      if (handUsed.equals("R")) {
        rightFingerPosition = moveFingerPosition(number);
      }
    }

    return result;
  }

  public String selectFinger(int number,
                             int[] leftFingerPosition,
                             int[] rightFingerPosition,
                             String handUse) {
    String handUsed = switch (number) {
      case 1, 4, 7 -> "L";
      case 3, 6, 9 -> "R";
      case 2, 5, 8, 0 ->
          chooseFinger(number, leftFingerPosition, rightFingerPosition, handUse);
      default -> "";
    };

    return handUsed;
  }

  public String chooseFinger(int number,
                             int[] leftFingerPosition,
                             int[] rightFingerPosition,
                             String handUse) {
    int[] numberPosition = checkNumberPosition(number);

    int distanceBetweenLeftAndNumber
        = checkDistance(numberPosition, leftFingerPosition);
    int distanceBetweenRightAndNumber
        = checkDistance(numberPosition, rightFingerPosition);

    if (distanceBetweenLeftAndNumber != distanceBetweenRightAndNumber) {
      return chooseShortDistance(
          distanceBetweenLeftAndNumber, distanceBetweenRightAndNumber);
    }

    return chooseHandUse(handUse);
  }

  public int[] checkNumberPosition(int number) {
    return switch (number) {
      case 1 -> new int[]{0, 0};
      case 2 -> new int[]{0, 1};
      case 3 -> new int[]{0, 2};
      case 4 -> new int[]{1, 0};
      case 5 -> new int[]{1, 1};
      case 6 -> new int[]{1, 2};
      case 7 -> new int[]{2, 0};
      case 8 -> new int[]{2, 1};
      case 9 -> new int[]{2, 2};
      case 0 -> new int[]{3, 1};
      default -> new int[]{};
    };
  }

  public int checkDistance(int[] numberPosition, int[] fingerPosition) {
    int numberX = numberPosition[0];
    int numberY = numberPosition[1];
    int fingerX = fingerPosition[0];
    int fingerY = fingerPosition[1];

    return Math.abs(numberX - fingerX) + Math.abs(numberY - fingerY);
  }

  public String chooseShortDistance(
      int distanceBetweenLeftAndNumber, int distanceBetweenRightAndNumber) {
    return distanceBetweenLeftAndNumber < distanceBetweenRightAndNumber
        ? "L"
        : "R";
  }

  public String chooseHandUse(String handUse) {
    return handUse.equals("left")
        ? "L"
        : "R";
  }

  public int[] moveFingerPosition(int number) {
    return checkNumberPosition(number);
  }
}

File: 220916/JavaScript/key-pad.js
function getDistance(locatedNumber, target) {
  const keyPad = {
    1: [0, 0],
    2: [0, 1],
    3: [0, 2],
    4: [1, 0],
    5: [1, 1],
    6: [1, 2],
    7: [2, 0],
    8: [2, 1],
    9: [2, 2],
    '*': [3, 0],
    0: [3, 1],
    '#': [3, 2],
  };

  const currentPosition = keyPad[locatedNumber];
  const targetPosition = keyPad[target];

  return Math.abs(targetPosition[0] - currentPosition[0])
    + Math.abs(targetPosition[1] - currentPosition[1]);
}

export default function solution(numbers, handUse) {
  const answer = [];

  const handPosition = {
    leftHand: '*',
    rightHand: '#',
  };

  numbers.forEach((number) => {
    if (number === 1 || number === 4 || number === 7) {
      answer.push('L');
      handPosition.leftHand = number;
      return;
    }

    if (number === 3 || number === 6 || number === 9) {
      answer.push('R');
      handPosition.rightHand = number;
      return;
    }

    const leftHandDistance = getDistance(handPosition.leftHand, number);
    const rightHandDistance = getDistance(handPosition.rightHand, number);

    if (leftHandDistance > rightHandDistance) {
      answer.push('R');
      handPosition.rightHand = number;
      return;
    }

    if (leftHandDistance < rightHandDistance) {
      answer.push('L');
      handPosition.leftHand = number;
      return;
    }

    if (leftHandDistance === rightHandDistance) {
      if (handUse === 'right') {
        answer.push('R');
        handPosition.rightHand = number;
        return;
      }

      if (handUse === 'left') {
        answer.push('L');
        handPosition.leftHand = number;
      }
    }
  });

  return answer.join('');
}
